front/src/applications/operationalStudies/components/MacroEditor/utils.ts

```diff
@@ -244,29 +248,98 @@ export const getTrainrunFrequencyFromTimetableItem = (
+      colorDarkModeMuted: CATEGORY_COLOR_VARIANTS[`${baseColor}10`],
+      colorDarkModeRelated: CATEGORY_COLOR_VARIANTS[`${baseColor}50`],
+    })
+  );
+
+  const subCategoryPage = await dispatch(
+    osrdEditoastApi.endpoints.getSubCategory.initiate({ page: 1 }, { subscribe: false })
+  ).unwrap();
 
-export const getTrainMainCategoryFromId = (id: number): TrainMainCategory | null => {
-  for (const key of OSRD_TRAINRUN_CATEGORY_MAPPING.keys()) {
-    if (OSRD_TRAINRUN_CATEGORY_MAPPING.get(key)!.id === id) {
-      return key !== 'NO_CATEGORY' ? key : null;
+  const subCategories: SubCategory[] = [...subCategoryPage.results];
+
+  const subCategoriesFromTrainCategoryToNGE = categories.filter(
+    (_, i) => i > TRAIN_MAIN_CATEGORY_TO_NGE.length - 1
+  );
+
+  const subCategoriesFormatted = subCategoriesFromTrainCategoryToNGE.map((cat) => {
+    const subCategory = subCategories.find((subCat) => subCat.code === cat.colorRef);
+    if (!subCategory) {
+      throw new Error(`SubCategory with code ${cat.colorRef} not found.`);
     }
+    return {
+      ...subCategory,
+      colorRef: subCategory.code,
+      color: subCategory.color,
+      hovered_color: subCategory.hovered_color,
+      background_color: subCategory.background_color,
+    };
+  });
+
+  const subColors = subCategoriesFormatted.map((cat, index) => ({
+    id: mainColors.length + index,
+    colorRef: cat.colorRef,
+    color: cat.color,
+    colorFocus: cat.hovered_color,
+    colorMuted: cat.background_color,
+    colorRelated: cat.color,
+    colorDarkMode: cat.color,
+    colorDarkModeFocus: cat.hovered_color,
+    colorDarkModeMuted: cat.background_color,
+    colorDarkModeRelated: cat.color,
+  }));
+
+  return [...mainColors, ...subColors];
+};
+
+export const getTrainrunCategoryCodeFromId = (
+  trainrunCategories: TrainrunCategory[],
+  id: number
+): (TrainMainCategory | SubCategory['code']) | null => {
+  const category = trainrunCategories.find((cat) => cat.id === id);
+  if (!category) return null;
+  if (category.id < TRAIN_MAIN_CATEGORY_TO_NGE.length) {
+    return TRAIN_MAIN_CATEGORY_TO_NGE[category.id].trainCategory;
   }
-  throw new Error(`Unknown train category id: ${id}`);
+  return category.colorRef;
 };
 
 // If the categoryKey is null or undefined, we return the ID for 'NO_CATEGORY'
-export const getTrainrunCategoryId = (categoryKey?: TrainMainCategory | null) =>
-  OSRD_TRAINRUN_CATEGORY_MAPPING.get(categoryKey || 'NO_CATEGORY')!.id;
+export const getTrainrunCategoryId = (
+  trainrunCategories: TrainrunCategory[],
+  category?: TrainCategory | null
+): number => {
+  if (category && isMainCategory(category)) {
+    const mainKey = category.main_category;
+    return OSRD_TRAINRUN_MAIN_CATEGORY_CODE_MAPPING.get(mainKey)!.id;
+  }
+
+  const subKey = category!.sub_category_code;
+  return trainrunCategories.find((cat) => cat.colorRef === subKey)!.id;
+};
+
+export const buildCategory = (category: string | null) => {
```

Comment:
```suggestion
export const buildCategory = (trainrunCategoryCode: string | null) => {
```

Nit: same as naming comments

Reply:
Done